OemPkg/OemConfigPolicyCreatorPei/OemConfigPolicyCreatorPei.c

```diff
@@ -0,0 +1,256 @@
+  UINTN                            UnicodeNameSize = 0;
+  // get a buffer of the max name size
+  CHAR16                           UnicodeName[CONF_VAR_NAME_LEN];
+  CONFIG_VAR_LIST_ENTRY            VarListEntry;
+  UINTN                            VarListSize;
+  VOID                             *ConfListPtr;
+
+  Status = PeiServicesLocatePpi (
+             &gEfiPeiReadOnlyVariable2PpiGuid,
+             0,
+             NULL,
+             (VOID **)&PPIVariableServices
+             );
+
+  if (EFI_ERROR (Status)) {
+    DEBUG ((DEBUG_ERROR, "%a failed to locate variable services!\n", __FUNCTION__));
+    ASSERT (FALSE);
+    return EFI_NOT_READY;
+  }
+
+  // first figure out how much space we need to allocate for the ConfPolicy
+  for (i = 0; i < NumKnobs; i++) {
+    // the var list will use the Unicode version of the name, gKnobData has the ASCII version
+    NeededSize += VAR_LIST_SIZE(gKnobData[i].NameSize * 2, gKnobData[i].ValueSize);
+    DEBUG ((DEBUG_ERROR, "OSDDEBUG NeededSize: %x\n", NeededSize));
+  }
+
+  if (NeededSize > MAX_UINT16) {
+    DEBUG((DEBUG_ERROR, "%a config is greater than 64k! Too large for what policy service supports\n", __FUNCTION__));
+    ASSERT (FALSE);
+    return EFI_UNSUPPORTED;
+  }
+
+  *ConfPolicy = AllocatePool (NeededSize);
+
+  if (*ConfPolicy == NULL) {
+    DEBUG ((DEBUG_ERROR, "%a failed to allocate Conf Policy memory!\n", __FUNCTION__));
+    ASSERT (FALSE);
+    return EFI_OUT_OF_RESOURCES;
+  } 
+
+  *ConfPolicySize = (UINT16)NeededSize;
+
+  // now go through and populate the Conf Policy
+  for (i = 0; i < NumKnobs; i++) {
+    UnicodeNameSize = gKnobData[i].NameSize * 2;
+    AsciiStrToUnicodeStrS (gKnobData[i].Name, UnicodeName, gKnobData[i].NameSize);
+
+    VarListEntry.Name = UnicodeName;
+    VarListEntry.Guid = gKnobData[i].VendorNamespace;
+    // hardcoded for now
+    VarListEntry.Attributes = 7;
+    VarListEntry.Data = gKnobData[i].CacheValueAddress;
+    VarListEntry.DataSize = gKnobData[i].ValueSize;
+
+    DUMP_HEX(DEBUG_ERROR, 0, gKnobData[i].CacheValueAddress, sizeof(BOOLEAN), "OSDDEBUG: ");
+
+    // check if value has been overridden in variable storage
+    VarSize = gKnobData[i].ValueSize;
+    Status  = PPIVariableServices->GetVariable (
+                                     PPIVariableServices,
+                                     UnicodeName,
+                                     &gKnobData[i].VendorNamespace,
+                                     NULL,
+                                     &VarSize,
+                                     gKnobData[i].CacheValueAddress
+                                     );
+
+    // if variable services fails other than failing to find the variable (it was not overidden)
+    // or a size mismatch (stale data from a previous definition of the knob), we should fail
+    // as we may be missing overidden knobs
+    if (EFI_ERROR (Status) && (Status != EFI_NOT_FOUND) && (Status != EFI_BUFFER_TOO_SMALL)) {

```

Comment:
Do we really proceed to conversion to variable list if buffer is too small?

Reply:
This was missing a commit that changed this logic significantly, please see latest.